NebulaModel/DataStructures/ToggleSwitch.cs

```diff
@@ -0,0 +1,33 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+
+namespace NebulaModel.DataStructures
+{
+    public sealed class ToggleSwitch
+    {
+        int onCount;
+
+        public static implicit operator bool(ToggleSwitch toggle) => toggle.onCount > 0;
```

Comment:
This now handles re-entrant states. Probably not required, but using Interlocked this is now also thread-safe.